File: ProjectTracker20vb/PTWin/ResourceEdit.vb
Imports ProjectTracker.Library

Public Class ResourceEdit
  Inherits WinPart

  Private WithEvents mResource As Resource

  Public ReadOnly Property Resource() As Resource
    Get
      Return mResource
    End Get
  End Property

  Public Sub New(ByVal resource As Resource)

    InitializeComponent()

    mResource = resource
    mResource.BeginEdit()

    Me.RoleListBindingSource.DataSource = RoleList.GetList
    Me.ResourceBindingSource.DataSource = mResource

    ApplyAuthorizationRules()

  End Sub

#Region " WinPart Code "

  Protected Overrides Function GetIdValue() As Object

    Return mResource

  End Function

  Public Overrides Function ToString() As String

    Return mResource.FullName

  End Function

  Private Sub ResourceEdit_CurrentPrincipalChanged( _
  ByVal sender As Object, _
  ByVal e As System.EventArgs) _
  Handles Me.CurrentPrincipalChanged

    ApplyAuthorizationRules()

  End Sub

#End Region

  Private Sub ApplyAuthorizationRules()

    ' have the controls enable/disable/etc
    Me.ReadWriteAuthorization1.ResetControlAuthorization()

    Dim canEdit As Boolean = _
      ProjectTracker.Library.Project.CanEditObject

    ' enable/disable appropriate buttons
    Me.OKButton.Enabled = canEdit
    Me.ApplyButton.Enabled = canEdit
    Me.Cancel_Button.Enabled = canEdit

    ' enable/disable role column in grid
    Me.AssignmentsDataGridView.Columns(3).ReadOnly = Not canEdit

  End Sub

  Private Sub SaveResource(ByVal rebind As Boolean)

    Using busy As New StatusBusy("Saving...")
      ' stop the flow of events
      Me.ResourceBindingSource.RaiseListChangedEvents = False
      Me.AssignmentsBindingSource.RaiseListChangedEvents = False

      ' do the save
      Dim temp As Resource = mResource.Clone
      temp.ApplyEdit()
      Try
        mResource = temp.Save
        mResource.BeginEdit()
        If rebind Then
          ' rebind the UI
          Me.ResourceBindingSource.DataSource = Nothing
          Me.AssignmentsBindingSource.DataSource = Nothing
          Me.ResourceBindingSource.DataSource = mResource
          ApplyAuthorizationRules()
        End If

      Catch ex As Csla.DataPortalException
        MessageBox.Show(ex.BusinessException.ToString, _
          "Error saving", MessageBoxButtons.OK, _
          MessageBoxIcon.Exclamation)

      Catch ex As Exception
        MessageBox.Show(ex.ToString, _
          "Error saving", MessageBoxButtons.OK, _
          MessageBoxIcon.Exclamation)

      Finally
        Me.ResourceBindingSource.RaiseListChangedEvents = True
        Me.AssignmentsBindingSource.RaiseListChangedEvents = True
      End Try

    End Using

  End Sub

  Private Sub OKButton_Click(ByVal sender As System.Object, _
    ByVal e As System.EventArgs) Handles OKButton.Click

    SaveResource(False)
    Me.Close()

  End Sub

  Private Sub ApplyButton_Click(ByVal sender As System.Object, _
    ByVal e As System.EventArgs) Handles ApplyButton.Click

    SaveResource(True)

  End Sub

  Private Sub Cancel_Button_Click(ByVal sender As System.Object, _
    ByVal e As System.EventArgs) Handles Cancel_Button.Click

    mResource.CancelEdit()

  End Sub

  Private Sub CloseButton_Click(ByVal sender As System.Object, _
    ByVal e As System.EventArgs) Handles CloseButton.Click

    mResource.CancelEdit()
    Me.Close()

  End Sub

  Private Sub AssignButton_Click(ByVal sender As System.Object, _
    ByVal e As System.EventArgs) Handles AssignButton.Click

    Dim dlg As New ProjectSelect
    If dlg.ShowDialog = DialogResult.OK Then
      Try
        mResource.Assignments.AssignTo(dlg.ProjectId)

      Catch ex As InvalidOperationException
        MessageBox.Show(ex.ToString, _
          "Error assigning", MessageBoxButtons.OK, _
          MessageBoxIcon.Information)

      Catch ex As Exception
        MessageBox.Show(ex.ToString, _
          "Error assigning", MessageBoxButtons.OK, _
          MessageBoxIcon.Exclamation)
      End Try
    End If

  End Sub

  Private Sub UnassignButton_Click(ByVal sender As System.Object, _
    ByVal e As System.EventArgs) Handles UnassignButton.Click

    If Me.AssignmentsDataGridView.SelectedRows.Count > 0 Then
      Dim projectId As Guid = _
        CType(Me.AssignmentsDataGridView.SelectedRows(0).Cells(0).Value, Guid)
      mResource.Assignments.Remove(projectId)
    End If

  End Sub

  Private Sub AssignmentsDataGridView_CellContentClick(ByVal sender As System.Object, _
    ByVal e As System.Windows.Forms.DataGridViewCellEventArgs) _
    Handles AssignmentsDataGridView.CellContentClick

    If e.ColumnIndex = 1 And e.RowIndex > -1 Then
      Dim projectId As Guid = _
        CType(Me.AssignmentsDataGridView.Rows(e.RowIndex).Cells(0).Value, Guid)
      MainForm.ShowEditProject(projectId)
    End If

  End Sub

End Class

File: ProjectTracker20cs/PTWin/ResourceEdit.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;
using ProjectTracker.Library;

namespace PTWin
{
  public partial class ResourceEdit : WinPart
  {

    private Resource _resource;

    public Resource Resource
    {
      get { return _resource; }
    }

    public ResourceEdit(Resource resource)
    {
      InitializeComponent();

      _resource = resource;

      this.CurrentPrincipalChanged += new EventHandler(ResourceEdit_CurrentPrincipalChanged);
      _resource.PropertyChanged += new PropertyChangedEventHandler(mResource_PropertyChanged);

      this.RoleListBindingSource.DataSource = RoleList.GetList();
      this.ResourceBindingSource.DataSource = _resource;

      ApplyAuthorizationRules();
    }

    #region WinPart Code

    protected internal override object GetIdValue()
    {
      return _resource;
    }

    public override string ToString()
    {
      return _resource.FullName;
    }

    private void ResourceEdit_CurrentPrincipalChanged(object sender, EventArgs e)
    {
      ApplyAuthorizationRules();
    }

    #endregion

    private void ApplyAuthorizationRules()
    {
      // have the controls enable/disable/etc
      this.ReadWriteAuthorization1.ResetControlAuthorization();

      bool canEdit = ProjectTracker.Library.Project.CanEditObject();

      // enable/disable appropriate buttons
      this.OKButton.Enabled = canEdit;
      this.ApplyButton.Enabled = canEdit;
      this.Cancel_Button.Enabled = canEdit;

      // enable/disable role column in grid
      this.AssignmentsDataGridView.Columns[3].ReadOnly = !canEdit;
    }

    private void SaveResource(bool rebind)
    {
      using (StatusBusy busy = new StatusBusy("Saving..."))
      {
        // stop the flow of events
        this.ResourceBindingSource.RaiseListChangedEvents = false;
        this.AssignmentsBindingSource.RaiseListChangedEvents = false;
        // do the save
        this.ResourceBindingSource.EndEdit();
        this.AssignmentsBindingSource.EndEdit();
        try
        {
          Resource temp = Resource.Clone();
          _resource = temp.Save();
          if (rebind)
          {
            // rebind the UI
            this.ResourceBindingSource.DataSource = null;
            this.AssignmentsBindingSource.DataSource = this.ResourceBindingSource;
            this.ResourceBindingSource.DataSource = _resource;
            ApplyAuthorizationRules();
          }
        }
        catch (Csla.DataPortalException ex)
        {
          MessageBox.Show(ex.BusinessException.ToString(),
            "Error saving", MessageBoxButtons.OK,
            MessageBoxIcon.Exclamation);
        }
        catch (Exception ex)
        {
          MessageBox.Show(ex.ToString(),
            "Error Saving", MessageBoxButtons.OK,
            MessageBoxIcon.Exclamation);
        }
        finally
        {
          this.ResourceBindingSource.RaiseListChangedEvents = true;
          this.AssignmentsBindingSource.RaiseListChangedEvents = true;
          this.ResourceBindingSource.ResetBindings(false);
          this.AssignmentsBindingSource.ResetBindings(false);
        }
      }
    }

    private void OKButton_Click(object sender, EventArgs e)
    {
      SaveResource(false);
      this.Close();
    }

    private void ApplyButton_Click(object sender, EventArgs e)
    {
      SaveResource(true);
    }

    private void Cancel_Button_Click(object sender, EventArgs e)
    {
      _resource.CancelEdit();
    }

    private void CloseButton_Click(object sender, EventArgs e)
    {
      _resource.CancelEdit();
      this.Close();
    }

    private void AssignButton_Click(object sender, EventArgs e)
    {
      ProjectSelect dlg = new ProjectSelect();
      if (dlg.ShowDialog() == DialogResult.OK)
        try
        {
          _resource.Assignments.AssignTo(dlg.ProjectId);
        }
        catch (InvalidOperationException ex)
        {
          MessageBox.Show(ex.ToString(),
            "Error Assigning", MessageBoxButtons.OK,
            MessageBoxIcon.Information);
        }
        catch (Exception ex)
        {
          MessageBox.Show(ex.ToString(),
            "Error Assigning", MessageBoxButtons.OK,
            MessageBoxIcon.Exclamation);
        }
    }

    private void UnassignButton_Click(object sender, EventArgs e)
    {
      if (this.AssignmentsDataGridView.SelectedRows.Count > 0)
      {
        Guid projectId = (Guid)this.AssignmentsDataGridView.SelectedRows[0].Cells[0].Value;
        _resource.Assignments.Remove(projectId);
      }
    }

    private void mResource_PropertyChanged(object sender, PropertyChangedEventArgs e)
    {
      if (e.PropertyName == "IsDirty")
      {
        this.ResourceBindingSource.ResetBindings(true);
        this.AssignmentsBindingSource.ResetBindings(true);
      }
    }

    private void AssignmentsDataGridView_CellContentClick(object sender, DataGridViewCellEventArgs e)
    {
      if (e.ColumnIndex == 1 && e.RowIndex > -1)
      {
        Guid projectId = (Guid)this.AssignmentsDataGridView.Rows[e.RowIndex].Cells[0].Value;
        MainForm.Instance.ShowEditProject(projectId);
      }
    }
  }
}
